### ISSUE TITLE
There may be a bug in the ConfusionMatrix class.

### ISSUE THREAD

**Issue by yuzhj** [ultralytics/ultralytics#18407]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

There exists a pred_box A and a gt_box B, but the IoU of A and B is 0. In this case, the FNs will increase by 1, but the FPs will not.
![image](https://github.com/user-attachments/assets/d8fef049-57f8-4552-a4bc-67afad0acd2c)

### Environment

。。。

### Minimal Reproducible Example

。。。

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @yuzhj, thank you for your interest in Ultralytics 🚀! We appreciate your detailed report regarding the `ConfusionMatrix` class bug with IoU behavior.

If this is a 🐛 Bug Report, we kindly ask you to provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and debug the issue. The example should clearly demonstrate the behavior you described, ideally with code snippets or data necessary to replicate the scenario. This will enable our team to investigate more efficiently.

We also recommend upgrading to the latest `ultralytics` package along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify if the issue persists in the most recent version:

```bash
pip install -U ultralytics
```

For more help, feel free to explore the [Docs](https://docs.ultralytics.com) which include [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage as well as debugging resources. If this is a custom implementation issue or code-level question, additional context (e.g., input data, expected output, and observed behavior) could also be helpful.

## Community and Support

You're welcome to join the Ultralytics community where it suits you best:
- Real-time chat? 📡 [Discord](https://discord.com/invite/ultralytics)
- In-depth discussions? 💬 [Discourse](https://community.ultralytics.com)
- Engage and share knowledge? 🎯 [Subreddit](https://reddit.com/r/Ultralytics)

## Verified Environments

YOLO can be run in the following thoroughly tested environments, complete with up-to-date dependencies ([CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), [PyTorch](https://pytorch.org/)):

- **Notebooks** with GPU support: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: See the [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If the above badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing on macOS, Windows, and Ubuntu. These tests include validation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/).

This is an automated response to help you get started, but rest assured, an Ultralytics engineer will review and assist with your issue shortly. Thank you for your patience and for using Ultralytics! 😊

**Comment by Y-T-G** [comment]
Can you create a minimum reproducible example?

**Comment by yuzhj** [comment]
> Can you create a minimum reproducible example?

yes
```python
from ultralytics.utils.metrics import ConfusionMatrix
import torch

matrix = ConfusionMatrix(nc=1)
detections = torch.tensor([[50, 50, 100, 100, 0.6, 0]])
gt_bboxes = torch.tensor([[200, 200, 300, 300]])
gt_cls = torch.tensor([0])

matrix.process_batch(detections, gt_bboxes, gt_cls)

print(matrix.matrix)
```

**Comment by Y-T-G** [comment]
Thanks. Can you create a PR with the fix?

**Comment by yuzhj** [comment]
> Thanks. Can you create a PR with the fix?

OK

**Comment by glenn-jocher** [comment]
@yuzhj PR merged, thank you for your contributions!
